src/server/engine_shard_set.cc

```diff
@@ -56,6 +80,72 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+// This function checks 3 things:
+// 1. Don't try memory fragmentation if we don't use "enough" memory (control by
+// mem_defrag_threshold flag)
+// 2. That we had change in memory usage - to prevent endless loop
+// 3. in case the above is OK, make sure that we have a "gap" between usage and commited memory
+// (control by commit_use_threshold flag)
+bool EngineShard::DefragTaskArgs::IsRequired() {
+  static const uint64_t mem_size = max_memory_limit;
+  static const float threshold_mem = mem_size / GetFlag(FLAGS_mem_defrag_threshold);
+  static const float commit_use_threshold = GetFlag(FLAGS_commit_use_threshold);
+
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  // we want to make sure that we are not running this to many times - i.e.
+  // if there was no change to the memory, don't run this
+  if (MemUsageChanged(current_mem_commit, commited) ||
+      MemUsageChanged(current_mem_used, mem_in_use)) {
+    if (threshold_mem < commited && mem_in_use != 0 &&
+        (float(mem_in_use * commit_use_threshold) < commited)) {
+      // we have way more commited then actual usage
+      current_mem_commit = commited;
+      current_mem_used = mem_in_use;
+      return true;
+    }
+  }
+  return false;
+}
+
+// for now this does nothing
+bool EngineShard::DoDefrag() {
+  // TODO: impl!
+  // Placeholder for now!
+  return defrag_args_.cursor > 0;
+}
+
+void EngineShard::DefragTaskArgs::Init() {
+  current_mem_commit = GetMallocCurrentCommitted();
+  current_mem_used = used_mem_current.load(memory_order_relaxed);
+  cursor = 0u;
+}
+
+// the memory de-fragmenting task is as follow:
+//  1. Check if memory usage is high enough
+//  2. Check if diff between commited and used memory is high enough
+//  3. Check if we have memory changes (to ensure that we not running endlessly).
+//  4. if all the above pass -> run on the shard and try to de-fragment memory by re-allocating
+//  values
+//     if the cursor for this is signal that we are not done, schedule the task to run at high
+//     priority otherwise lower the task priority so that it would not use the CPU when not required
+uint32_t EngineShard::DefragTask() {
+  const auto shard_id = db_slice().shard_id();
+  bool required_state = (defrag_args_.cursor == 0) && defrag_args_.IsRequired();
```

Comment:
should not it be
`required_state =  (defrag_args_.cursor != 0)  || defrag_args_.IsRequired()`  ? Do I misunderstand the meaning of this variable?

Reply:
yes - I already change this but since then I made some other changes with the patch so I didn't have time to commit new changes